# HG changeset patch
# User Rik <rik@octave.org>
# Date 1531543610 25200
#      Fri Jul 13 21:46:50 2018 -0700
# Node ID b3f6443f6b20daa1c16157d0985bc508388bd6e2
# Parent  d88bc4983ce7f9f79577351bff27d1d40712aea0
isdiag.m: Add fast path for detecting complex diagonal matrices.

* isdiag.m: Use strfind, not strcmp, to search typeinfo of input for
"diagonal matrix".  Add a BIST test to check class single inputs.

diff --git a/scripts/linear-algebra/isdiag.m b/scripts/linear-algebra/isdiag.m
--- a/scripts/linear-algebra/isdiag.m
+++ b/scripts/linear-algebra/isdiag.m
@@ -25,17 +25,17 @@
 ## Author: Massimiliano Fasi
 
 function retval = isdiag (A)
 
   if (nargin != 1)
     print_usage ();
   endif
 
-  if (strcmp (typeinfo (A), "diagonal matrix"))
+  if (strfind (typeinfo (A), "diagonal matrix"))
     retval = true;
   elseif ((isnumeric (A) || islogical (A)) && ndims (A) == 2)
     [i, j] = find (A);
     retval = all (i == j);
   else
     retval = false;
   endif
 
@@ -44,15 +44,16 @@ endfunction
 
 %!assert (isdiag ("string"), false)
 %!assert (isdiag (zeros (2,2,2)), false)
 %!assert (isdiag ([]))
 %!assert (isdiag (1))
 %!assert (isdiag ([1, 1]), false)
 %!assert (isdiag ([1; 1]), false)
 %!assert (isdiag (eye (10)))
+%!assert (isdiag (single (eye (10))))
 %!assert (isdiag (logical (eye (10))))
-%!assert (isdiag (speye (1e6)))
+%!assert (isdiag (speye (1e2)))
 %!assert (isdiag (diag (1:10)))
 
 ## Test input validation
 %!error isdiag ()
 %!error isdiag (1,2)
